### ISSUE TITLE
Document recommended virtualization providers and their strengths / weaknesess

### ISSUE THREAD

**Issue by deviantintegral** [laravel/homestead#1186]
I really appreciate that Homestead ships images for many different providers, especially given the historical difficulties of running multiple hypervisors or VM kernel modules at the same time.

[The introduction](https://laravel.com/docs/5.8/homestead#installation-and-setup) has one note about Hyper-V limitations, but no other guidance on recommended providers.

Personally, I was burned a few years back after paying for both VMWare Fusion and the Vagrant plugin. There was a bug in the Linux guest additions for Ubuntu that made it unusable (https://github.com/hashicorp/vagrant/issues/6775 I think) and VMWare took months to fix. I moved back to Virtualbox and haven't changed since.

It would be great if we could document some basic guidance on choosing a provider, especially if you have more than one to choose from. If there are usage stats for each provider, that could be a useful signal for new users as well. Or, if we don't have any guidance, we could link to Vagrant's [Default Provider](https://www.vagrantup.com/docs/providers/default.html) docs.

**Comment by svpernova09** [comment]
I wish Vagrant Cloud gave us stats by provider, they only give us total downloads :/

Once upon a time we only supported Virtualbox & VMware. Then dropped VMware when it became really problematic to build for. When I took over the project I started adding providers back as I could. VMware (with help from Taylor reimbursing me for licenses) was a lot of work, but since has become pretty "easy" to maintain. Parallels support came when they approached me about providing a free license to their software (which they still do, and I'm incredibly grateful for), and Hyper-V has been my own pet project because it's the fastest I've ever seen *any* Vagrant machine run.

Ranked by ease of use:
* Virtualbox
* Parallels
* VMware
* Hyper-V

Ranked by my own guess based on how much support I do:
* Virtualbox
* VMware
* Hyper-V
* Parallels

**Comment by svpernova09** [comment]
While on the subject, I'd LOVE to see someone do some benchmarks on folder / file sharing. The understood longtime "standard" has always been VMware -> Parallels -> Virtualbox -> NFS -> Samba but I've never really tested that.

**Comment by deviantintegral** [comment]
I've done a decent amount of FS benchmarking in the past using Drupal's CLI site install, which stat's a ton of files and won't have any persistent caches. I'll start work on scripting this.

**Comment by deviantintegral** [comment]
I'm giving a presentation tomorrow on Homestead, so I started writing a benchmark script in my example repository. See this branch, in particular `benchmark.sh`, `after.sh`, and the `providers` directory.

https://github.com/deviantintegral/homestead-d8-example/tree/benchmark-fs

I'm pretty impressed with how far along `vboxsf` has come. It used to be truly unusable. I may update the branch to add in more composer dependencies, since IME slowness comes from the number of shared files.

Here's the first set of results, macOS 10.14.5 / VirtualBox 6.0.8:

```csv
provider,sync_type,user,system,elapsed,cpu,inputs,outputs
virtualbox,nfs,7.59,2.19,0:53.17,18%,42912,616
virtualbox,nfs,7.88,2.12,0:46.12,21%,8,64
virtualbox,nfs,8.08,2.36,0:47.22,22%,0,64
virtualbox,rsync,7.36,1.04,0:44.22,19%,960,616
virtualbox,rsync,7.88,0.93,0:43.29,20%,1712,64
virtualbox,rsync,7.66,1.11,0:44.23,19%,688,64
virtualbox,smb,7.55,10.28,1:52.52,15%,608,616
virtualbox,smb,6.82,9.54,1:30.54,18%,0,64
virtualbox,smb,7.16,8.75,1:15.53,21%,0,64
virtualbox,vboxsf,6.98,23.57,1:15.26,40%,608,616
virtualbox,vboxsf,6.40,23.25,1:10.31,42%,0,64
virtualbox,vboxsf,6.20,22.02,1:07.85,41%,0,64
```

Which gives initial results as:

Sync type | Average of elapsed
----------- | ----------------
nfs | 00:00:49
rsync | 00:00:44
smb | 00:01:33
vboxsf | 00:01:11
Grand Total | 00:01:04

<img width="515" alt="Screen Shot 2019-06-19 at 7 32 16 PM" src="https://user-images.githubusercontent.com/255023/59808146-01ee7f80-92c9-11e9-8b45-f2fd5b5429d8.png">

**Comment by cbj4074** [comment]
I can say unequivocally (but only anecdotally -- I'll work on some "hard numbers") that Hyper-V + SMB is *much, much* faster than VirtualBox + SMB, and *much* faster even than VirtualBox + NFS.

In fact, I switched from VirtualBox + NFS to Hyper-V + SMB for this reason alone, and haven't looked back since.

**Comment by svpernova09** [comment]
Hyper-V + SMB is still the _fastest_ I've ever seen Vagrant (let alone Homestead) run. So much so I almost converted my daily work system to Windows.

**Comment by svpernova09** [comment]
It bombed on SMB, but here is what was recorded:

```
provider,sync_type,user,system,elapsed,cpu,inputs,outputs
virtualbox,nfs,8.99,3.14,1:00.49,20%,42912,696
virtualbox,nfs,7.75,2.24,0:45.97,21%,64,64
virtualbox,nfs,7.40,2.05,0:44.29,21%,0,64
virtualbox,rsync,6.06,0.87,0:36.36,19%,608,616
virtualbox,rsync,7.06,0.92,0:38.25,20%,392,64
virtualbox,rsync,7.18,0.94,0:40.23,20%,0,64
```

MacOS 10.14.5 (Mojave)
2.9Ghz Intel Core i7
16GB 2133 MHz LPDDR3
APPLE SSD SM0512L

**Comment by svpernova09** [comment]
@vongrippen What would be the best way to adapt/add Hyper-V, and even VMware testing to these? I assume we couldn't test Virtualbox + Hyper-V / VMware on the same system, but what are the share providers named on those so we can run a baseline on those systems?

**Comment by vongrippen** [comment]
Hmm, as long as Hyper-V is running you can't run other hypervisors. That does complicate testing by making it so that you'll have to run the Hyper-V tests separately, though you'll need to run the Parallels tests separately as well if you want to test that.

Pro-tip though, latest Virtualbox (6.0+) "can" run alongside Hyper-V but it actually uses the Windows Hypervisor when that happens instead of using Virtualbox's hypervisor. I'd honestly be curious to see if that had any affect on the VirtualBox results.

As for the providers on Hyper-V and VMware... On Hyper-V it's just SMB. Microsoft is super straightforward about that, it's SMB but essentially implemented in memory. On VMware, it's `vmhgfs`.

**Comment by svpernova09** [comment]
Gotcha, thanks.

**Comment by deviantintegral** [comment]
I have Windows 10 Pro in Bootcamp on my laptop, so I could actually do an identical-hardware test with Hyper-V. The only difference is that macOS is under FileVault, and W10 isn't encrypted at all (because Bitlocker isn't supported under Bootcamp).

**Comment by svpernova09** [comment]
@deviantintegral I'd be interested to see those results.

I don't have VMware on any native hardware atm. So I can't really do meaningful benchmarks there. I could do Virtualbox on Windows 10, and Hyper-V on Windows 10 (different hardware, different specs) I think the key is to make sure whoever is running the benchmarks needs to share software versions and hardware specs.

**Comment by browner12** [comment]
provider | sync_type | user | system | elapsed | cpu | inputs | outputs
-- | -- | -- | -- | -- | -- | -- | --
virtualbox | nfs | 7.25 | 1.71 | 01:12.6 | 12% | 42912 | 696
virtualbox | nfs | 6.4 | 1.7 | 00:55.9 | 14% | 64 | 64
virtualbox | nfs | 6.4 | 1.62 | 00:52.8 | 15% | 0 | 64
virtualbox | rsync | 5.86 | 0.78 | 00:42.9 | 15% | 608 | 616
virtualbox | rsync | 6.02 | 0.71 | 00:41.7 | 16% | 400 | 64
virtualbox | rsync | 5.83 | 0.84 | 00:39.8 | 16% | 0 | 64
virtualbox | vboxsf | 6.42 | 23.97 | 02:09.9 | 23% | 608 | 616
virtualbox | vboxsf | 5.46 | 21.82 | 01:18.8 | 34% | 0 | 64
virtualbox | vboxsf | 5.73 | 22.28 | 01:20.3 | 34% | 0 | 64

MacOS Mojave (10.14.5)
3.4 GHz Intel Core i5
24 GB 1600 MHz DDR3
Nvidia GeForce GTX 775M 2GB
1 TB HDD
Vagrant 2.2.5
Virtualbox 6.0.6

**Comment by cbj4074** [comment]
I don't know if it's helpful in the context of this conversation, but it's possible to disable Hyper-V only temporarily upon restarting Windows, which enables one to test using Hyper-V and other hypervisors on the same physical machine, if so desired.

The best technique that I've seen is described in https://www.hanselman.com/blog/SwitchEasilyBetweenVirtualBoxAndHyperVWithABCDEditBootEntryInWindows81.aspx .

Basically, one creates a new boot entry using `bcdedit` that includes a switch to disable Hyper-V. Then, whenever one wishes to reboot with Hyper-V disabled, the process is as simple as holding-down the Shift key when clicking `Restart` (in the Start menu), which prompts the user for the boot entry to use upon rebooting.

**Comment by deviantintegral** [comment]
I just finished benchmarking on the same laptop I used for the macOS benchmarks with hyper-v:

| provider | sync_type | user  | system | elapsed | cpu | inputs | outputs | 
|----------|-----------|-------|--------|---------|-----|--------|---------| 
| hyperv   | smb       | 12.22 | 4.62   | 01:15.5 | 22% | 608    | 616     | 
| hyperv   | smb       | 12.17 | 4.45   | 01:14.6 | 22% | 0      | 64      | 
| hyperv   | smb       | 12.32 | 4.31   | 01:14.1 | 22% | 0      | 64      | 
| hyperv   | rsync     | 8.45  | 1.4    | 00:49.0 | 20% | 55912  | 616     | 
| hyperv   | rsync     | 8.58  | 1.26   | 00:46.6 | 21% | 8      | 64      | 
| hyperv   | rsync     | 8.36  | 1.19   | 00:46.1 | 20% | 0      | 64      | 

Looks like for this test case windows + hyperv + smb is slower than macos + virtualbox + nfs. Ignoring IO with rsync gives pretty much identical results.

**Comment by svpernova09** [comment]
Calling this one done. May want to connect w/ @taylorotwell and see what the best way to incorporate some of these into the docs may be? Maybe better off as a blog post.
